### PR TITLE
Damage plasticity model

### PR DESCRIPTION
**Pull request by vprithiv**
https://github.com/idaholab/blackbear/pull/265

Continuing Amit's Work on damage plasticity model
ref https://github.com/idaholab/blackbear/issues/184

### CODE REVIEW

File: src/materials/ComputeMultipleInelasticDamageStress.C [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                             "DamagePlasticityStressUpdate"); 
+  return params;
+}
+
+ComputeMultipleInelasticDamageStress::ComputeMultipleInelasticDamageStress(
+    const InputParameters & parameters)
+  : ComputeMultipleInelasticStress(parameters),
+    _D(getMaterialProperty<Real>("damage_variable")),
+    _D_old(getMaterialPropertyOld<Real>("damage_variable"))
+{
```
Comment (bwspenc) [inline src/materials/ComputeMultipleInelasticDamageStress.C:34]: Add code like this to verify that `_models` contains only a `DamagePlasticityStressUpdate` model:
```
if (_models.size() != 1)
  paramError("ComputeMultipleInelasticDamageStress currently can only have one model specified in 'inelastic_models'");
  const DamagePlasticityStressUpdate * temp = dynamic_cast<DamagePlasticityStressUpdate *>(_models[0]);
  if (!temp)
    paramError("Model " + _models[0] + " is not a DamagePlasticityStressUpdate object");

File: src/materials/DamagePlasticityStressUpdate.C [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+    d_sqrt_2J2 = RankTwoTensor();
+  else
+    d_sqrt_2J2 = 0.5 * std::sqrt(2.0 / J2) *
+                 RankTwoTensor(stress_params[0], stress_params[1], stress_params[2], 0, 0, 0)
+                     .dsecondInvariant();
+
+  Real D = damageVar(stress_params, intnl);
+
+  for (unsigned int i = 0; i < _num_sp; ++i)
+    r[i] = (_alfa_p + d_sqrt_2J2(i, i)) * std::pow((1. - D), 1);
```
Comment (bwspenc) [inline src/materials/DamagePlasticityStressUpdate.C:321]: Remove pow call.

File: src/materials/DamagePlasticityStressUpdate.C [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+    _ft(getParam<Real>("yield_strength_in_tension")),
+    _FEt(getParam<Real>("fracture_energy_in_tension")),
+
+    _fyc(getParam<Real>("yield_strength_in_compression")),
+    _Dc(getParam<Real>("compressive_damage_at_max_compressive_strength")),
+    _fc(getParam<Real>("maximum_strength_in_compression")),
+    _FEc(getParam<Real>("fracture_energy_in_compression")),
+
+    _at(1.5 * std::sqrt(1 - _Chi) - 0.5),
+    _ac((2. * (_fc / _fyc) - 1. + 2. * std::sqrt(std::pow((_fc / _fyc), 2.) - _fc / _fyc))),
```
Comment (bwspenc) [inline src/materials/DamagePlasticityStressUpdate.C:97]: Integer power should use Utility::pow<2>()

File: src/materials/DamagePlasticityStressUpdate.C [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+    _sigma0(declareProperty<Real>("damaged_min_principal_stress")),
+    _sigma1(declareProperty<Real>("damaged_mid_principal_stress")),
+    _sigma2(declareProperty<Real>("damaged_max_principal_stress"))
+{
+}
+
+void
+DamagePlasticityStressUpdate::initQpStatefulProperties()
+{
+  // if (_current_elem->n_vertices() < 3)
```
Comment (bwspenc) [inline src/materials/DamagePlasticityStressUpdate.C:136]: Put a comment saying that there multiple potential ways to determine the correct element length, and these are some options that are still being considered.

File: include/materials/DamagePlasticityStressUpdate.h [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+#include "MultiParameterPlasticityStressUpdate.h"
+
+class DamagePlasticityStressUpdate : public MultiParameterPlasticityStressUpdate
+{
+public:
+  static InputParameters validParams();
+  DamagePlasticityStressUpdate(const InputParameters & parameters);
+
+  /**
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:25]: Delete these comments

File: include/materials/DamagePlasticityStressUpdate.h [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  MaterialProperty<Real> & _tD;
+  MaterialProperty<Real> & _cD;
+  MaterialProperty<Real> & _D;
+  MaterialProperty<Real> & _min_ep;
+  MaterialProperty<Real> & _mid_ep;
+  MaterialProperty<Real> & _max_ep;
+  MaterialProperty<Real> & _sigma0;
+  MaterialProperty<Real> & _sigma1;
+  MaterialProperty<Real> & _sigma2;
+
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:88]: Add doxygen comments for functions:
/**
 * Get the tensile strength
 * @param intnl Describe this parameter
 * @return tensile strength
 */

File: include/materials/DamagePlasticityStressUpdate.h [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+   * Does the model require the elasticity tensor to be isotropic?
+   */
+  bool requiresIsotropicTensor() override { return true; }
+
+protected:
+  virtual void initQpStatefulProperties() override;
+  virtual void finalizeReturnProcess(const RankTwoTensor & rotation_increment) override;
+
+private:
+  const Real _f_tol;
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:35]: Add one-liner doxygen descriptions of member variables
/// Yield function tolerance (user parameter)

File: src/materials/DamagePlasticityStressUpdate.C [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+}
+
+Real
+DamagePlasticityStressUpdate::ft(const std::vector<Real> & intnl) const
+{
+  Real sqrtPhi_t = std::sqrt(1. + _at * (2. + _at) * intnl[0]);
+  if (_zt > sqrtPhi_t / _at)
+    return _ft0 * std::pow(_zt - sqrtPhi_t / _at, (1. - _dt_bt)) * sqrtPhi_t;
+  else
+    return _ft0 * 1.E-6;
```
Comment (bwspenc) [inline src/materials/DamagePlasticityStressUpdate.C:538]: Look into this -- this doesn't seem quite right.

Comment (bwspenc) [reply src/materials/DamagePlasticityStressUpdate.C:538]: If it's a useful constant, do something like:
```
const tol = 1e-6;
return _ft0 * tol;

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: doc/content/source/materials/ComputeMultipleInelasticDamageStress.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Compute Multiple Inelastic Damage Stress
+
+!syntax description /Materials/ComputeMultipleInelasticDamageStress
+
+## Description
+
+`ComputeMultipleInelasticDamageStress` computes the damage stress.
+This ComputeMultipleInelasticStress is to be used with (/DamagePlasticityStressUpdate.md).
+## Example Input Files
```
[the remark this answers is not in the document]
Comment (vprithiv) [reply doc/content/source/materials/ComputeMultipleInelasticDamageStress.md:11]: Addressed

File: src/materials/ComputeMultipleInelasticDamageStress.C [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void
+ComputeMultipleInelasticDamageStress::computeQpJacobianMult()
+{
+  ComputeMultipleInelasticStress::computeQpJacobianMult();
+  _Jacobian_mult[_qp] = (1.0 - _D_old[_qp]) * _Jacobian_mult[_qp];
+  // _Jacobian_mult[_qp] = (1.0 - _D[_qp]) * _Jacobian_mult[_qp];
```
[the remark this answers is not in the document]
Comment (vprithiv) [reply src/materials/ComputeMultipleInelasticDamageStress.C]: Addressed

File: src/materials/ComputeMultipleInelasticDamageStress.C [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    RankTwoTensor & elastic_strain_increment,
+    RankTwoTensor & inelastic_strain_increment,
+    RankFourTensor & consistent_tangent_operator)
+{
+  _models[model_number]->updateState(elastic_strain_increment,
+                                     inelastic_strain_increment,
+                                     _rotation_increment[_qp],
+                                     _stress[_qp],
+                                     _stress_old[_qp] / (1.0 - _D_old[_qp]),
+                                     // _stress_old[_qp] / (1.0 - _D[_qp]),
```
[the remark this answers is not in the document]
Comment (vprithiv) [reply src/materials/ComputeMultipleInelasticDamageStress.C]: Addressed

File: doc/content/source/materials/ComputeMultipleInelasticDamageStress.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# Compute Multiple Inelastic Damage Stress
+
+!syntax description /Materials/ComputeMultipleInelasticDamageStress
+
+## Description
+
+`ComputeMultipleInelasticDamageStress` computes the damage stress.
+This ComputeMultipleInelasticStress is to be used with (/DamagePlasticityStressUpdate.md).
```
[the remark this answers is not in the document]
Comment (vprithiv) [reply doc/content/source/materials/ComputeMultipleInelasticDamageStress.md:10]: Addressed

File: src/materials/DamagePlasticityStressUpdate.C (hunk 4 above)
Comment (vprithiv) [reply src/materials/DamagePlasticityStressUpdate.C:139]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h (hunk 5 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h (hunk 6 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h]: Addressed ( Need to check InitializeVarV, Trial_stress_params, Stress_params)

File: include/materials/DamagePlasticityStressUpdate.h (hunk 7 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h]: Partially addressed (Some member variables are not understood)

File: src/materials/DamagePlasticityStressUpdate.C (hunk 2 above)
Comment (vprithiv) [reply src/materials/DamagePlasticityStressUpdate.C]: Addressed

File: src/materials/DamagePlasticityStressUpdate.C (hunk 3 above)
Comment (vprithiv) [reply src/materials/DamagePlasticityStressUpdate.C]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+  const Real _dPhic;
+  const Real _sqrtPhit_max;
+  const Real _sqrtPhic_max;
+  const Real _dt_bt;
+  const Real _dc_bc;
+  const Real _ft0;
+  const Real _fc0;
+  ///@}
+
+  /// Intermediate variable calcualted using  user parameter tip_smoother
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:74]: spelling

File: include/materials/DamagePlasticityStressUpdate.h [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+  const Real _small_smoother2;
+
+  const Real _sqrt3;
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:77]: Add 
/// square root of 3

File: include/materials/DamagePlasticityStressUpdate.h [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+  MaterialProperty<Real> & _gt;
+  ///fracture energy in compression
+  MaterialProperty<Real> & _gc;
+  ///tensile damage
+  MaterialProperty<Real> & _tD;
+  ///compression damage
+  MaterialProperty<Real> & _cD;
+  ///damage variable
+  MaterialProperty<Real> & _D;
+  ///minimum plastic strain
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:100]: eigenvalue...

File: include/materials/DamagePlasticityStressUpdate.h [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+  MaterialProperty<Real> & _max_ep;
+  ///damaged minimum principal stress
+  MaterialProperty<Real> & _sigma0;
+  ///damaged mid value of principal stress
+  MaterialProperty<Real> & _sigma1;
+  ///damaged maximum principal stress
+  MaterialProperty<Real> & _sigma2;
+
+  Real ft(const std::vector<Real> & intnl) const;
+  /**
```
Comment (bwspenc) [inline include/materials/DamagePlasticityStressUpdate.h:114]: Put all these above the function definitions. Consider adding '@return' for these too.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: include/materials/DamagePlasticityStressUpdate.h (hunk 13 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h:74]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h (hunk 14 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h:77]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h (hunk 15 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h:100]: Addressed

File: include/materials/DamagePlasticityStressUpdate.h (hunk 16 above)
Comment (vprithiv) [reply include/materials/DamagePlasticityStressUpdate.h:114]: Finished

File: src/materials/ComputeMultipleInelasticDamageStress.C (hunk 1 above)
Comment (vprithiv) [reply src/materials/ComputeMultipleInelasticDamageStress.C:35]: addressed

File: src/materials/DamagePlasticityStressUpdate.C (hunk 8 above)
Comment (bwspenc) [reply src/materials/DamagePlasticityStressUpdate.C:541]: This still needs to be addressed.

File: doc/content/bib/blackbear.bib [hunk 17]
```diff
@@ -244,3 +244,50 @@ @article{poyet2009temperature
...
+  author={Lee, Jeeho and Fenves, Gregory L},
+  journal={Journal of engineering mechanics},
+  volume={124},
+  number={8},
+  pages={892--900},
+  year={1998},
+  publisher={American Society of Civil Engineers}
+}
+
+@book{lee1996theory,
```
Comment (bwspenc) [inline doc/content/bib/blackbear.bib:259]: Change the following:
`@book` to `@phdthesis`
`@publisher` to `@school`

File: doc/content/source/materials/ComputeMultipleInelasticDamageStress.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
+# Compute Multiple Inelastic Damage Stress
+
+!syntax description /Materials/ComputeMultipleInelasticDamageStress
+
+## Description
+
+This class `ComputeMultipleInelasticDamageStress` computes the stress with the damage obtained from
```
Comment (bwspenc) [inline doc/content/source/materials/ComputeMultipleInelasticDamageStress.md:7]: I'd reword this whole section a bit:
```
`ComputeMultipleInelasticDamageStress` is a class based on `ComputeMultipleInelasticStress`
that is designed specifically to be used with [DamagePlasticityStressUpdate](DamagePlasticityStressUpdate.md).
The main difference between this class and the standard `ComputeMultipleInelasticStress` class is that
it makes some modifications to both the current and old stress to account for the damage in a way that
is appropriate for the damaged plasticity model. It divides the old stress passed to the plasticity model
by $(1-D)$, and multiplies the new stress by $(1-D)$, where $D$ is the damage index.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: doc/content/source/materials/ComputeMultipleInelasticDamageStress.md (hunk 18 above)
Comment (vprithiv) [reply doc/content/source/materials/ComputeMultipleInelasticDamageStress.md]: Addressed

File: doc/content/bib/blackbear.bib (hunk 17 above)
Comment (vprithiv) [reply doc/content/bib/blackbear.bib]: Edited

File: src/materials/DamagePlasticityStressUpdate.C (hunk 8 above)
Comment (vprithiv) [reply src/materials/DamagePlasticityStressUpdate.C]: Wrote a comment per discussion

**Review by bwspenc — APPROVED** [review]

**Review by bwspenc — APPROVED** [review]
